distributed/utils_test.py

```diff
@@ -1595,24 +1596,67 @@ def check_thread_leak():
             assert False, (bad_thread, call_stacks)
 
 
-@contextmanager
-def check_process_leak(check=True):
-    for proc in mp_context.active_children():
+def wait_active_children(timeout: float) -> list[multiprocessing.Process]:
+    """Wait until timeout for mp_context.active_children() to terminate.
+    Return list of active subprocesses after the timeout expired.
+    """
+    t0 = time()
+    while True:
+        # Do not sample the subprocesses once at the beginning with
+        # `for proc in mp_context.active_children: ...`, assume instead that new
+        # children processes may be spawned before the timeout expires.
+        children = mp_context.active_children()
+        if not children:
+            return []
+        join_timeout = timeout - time() + t0
+        if join_timeout <= 0:
+            return children
+        children[0].join(timeout=join_timeout)
+
+
+def term_or_kill_active_children(timeout: float) -> None:
+    """Send SIGTERM to mp_context.active_children(), wait up to 3 seconds for processes
+    to die, then send SIGKILL to the survivors
+    """
+    children = mp_context.active_children()
+    for proc in children:
         proc.terminate()
 
-    yield
+    children = wait_active_children(timeout=timeout)
+    for proc in children:
+        proc.kill()
 
-    if check:
-        for i in range(200):
-            if not set(mp_context.active_children()):
-                break
-            else:
-                sleep(0.2)
-        else:
-            assert not mp_context.active_children()
+    children = wait_active_children(timeout=30)
+    if children:  # pragma: nocover
+        logger.warning("Leaked unkillable children processes: %s", children)
+        # It should be impossible to ignore SIGKILL on Linux/MacOSX
+        assert WINDOWS
 
-    for proc in mp_context.active_children():
-        proc.terminate()
+
+@contextmanager
+def check_process_leak(
+    check: bool = True, check_timeout: float = 40, term_timeout: float = 3
+):
+    """Terminate any subprocesses spawned inside the context
```

Comment:
```suggestion
    """Terminate any currently-running subprocesses at both the beginning and end of this context
```
The fact that `term_or_kill_active_children` is also called at the beginning seems important to mention.